components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt

```diff
@@ -263,13 +266,148 @@ class GitpodConnectionProvider : GatewayConnectionProvider {
                     "${connectParams.gitpodHost}: ${connectParams.workspaceId}: failed to process workspace updates:",
                     t
                 )
-                errorMessage.text = " failed to process workspace updates ${t.message}"
+                setErrorMessage("failed to process workspace updates ${t.message}")
             }
         }
 
         return GitpodConnectionHandle(connectionLifetime, connectionPanel, connectParams);
     }
 
+    private suspend fun resolveJoinLink(
+        ideUrl: URL,
+        ownerToken: String,
+        connectParams: ConnectParams
+    ): String? {
+        var resolveJoinLinkUrl = "https://24000-${ideUrl.host}/joinLink"
+        if (!connectParams.backendPort.isNullOrBlank()) {
+            resolveJoinLinkUrl += "?backendPort=${connectParams.backendPort}"
+        }
+        return fetchWS(resolveJoinLinkUrl, connectParams, ownerToken)
+    }
+
+    private fun resolveCredentials(
+        ideUrl: URL,
+        userName: String,
+        password: String,
+        hostKeys: List<SSHHostKey>
+    ): RemoteCredentialsHolder {
+        val credentials = RemoteCredentialsHolder()
+        credentials.setHost(ideUrl.host)
+        credentials.port = 22
+        credentials.userName = userName
+        credentials.password = password
+        credentials.connectionBuilder().withSshConnectionConfig {
+            val hostKeyVerifier = it.hostKeyVerifier
+            if (hostKeyVerifier is OpenSshLikeHostKeyVerifier) {
+                val acceptHostKey = acceptHostKey(ideUrl, hostKeys)
+                it.copy(
+                    hostKeyVerifier = hostKeyVerifier.copy(
+                        acceptChangedHostKey = acceptHostKey,
+                        acceptUnknownHostKey = acceptHostKey
+                    )
+                )
+            } else {
+                it
+            }
+        }.connect()
+        return credentials
+    }
+
+    private suspend fun resolveHostKeys(
+        ideUrl: URL,
+        connectParams: ConnectParams
+    ): List<SSHHostKey>? {
+        val hostKeysValue =
+            fetchWS("https://${ideUrl.host}/_ssh/host_keys", connectParams, null)
+        if (hostKeysValue.isNullOrBlank()) {
+            return null
+        }
+        return with(jacksonMapper) {
+            propertyNamingStrategy = PropertyNamingStrategies.SnakeCaseStrategy()
+            readValue(hostKeysValue, object : TypeReference<List<SSHHostKey>>() {})
+        }
+    }
+
+    private fun acceptHostKey(
+        ideUrl: URL,
+        hostKeys: List<SSHHostKey>
+    ): AskAboutHostKey {
+        val hostKeysByType = hostKeys.groupBy({ it.type.lowercase() }) { it.hostKey }
+        val acceptHostKey: AskAboutHostKey = { hostName, keyType, fingerprint, _ ->
+            if (hostName != ideUrl.host) {
```

Comment:
@iQQBot Could you have a look please whether it makes sense to you?

Reply:
good catch, now yes

Reply:
Maybe we can print an URL in werft job somehow. We know a version and can query JB marketplace for URL, but not sure how to display it on results page.

Reply:
ok, let's wait for workspace team to deploy and then we can merge